ssl/ech/ech_internal.c

```diff
@@ -160,4 +217,940 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+                BIO_free(handbuf);
+                return 0;
+            }
+            if (PACKET_remaining(&pkt) > 0) {
+                if (ssl3_finish_mac(s, PACKET_data(&pkt), PACKET_remaining(&pkt)) == 0) {
+                    SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                    BIO_free(handbuf);
+                    return 0;
+                }
+            }
+            BIO_free(handbuf);
+        }
+    }
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ossl_ech_ptranscript(s, "ech_swaperoo, after");
+# endif
+    /* Declare victory! */
+    s->ext.ech.attempted = 1;
+    s->ext.ech.success = 1;
+    s->ext.ech.done = 1;
+    s->ext.ech.grease = OSSL_ECH_NOT_GREASE;
+    /* time to call an ECH callback, if there's one */
+    if (s->ext.ech.es != NULL && s->ext.ech.done == 1
+        && s->hello_retry_request != SSL_HRR_PENDING
+        && s->ext.ech.cb != NULL) {
+        char pstr[OSSL_ECH_PBUF_SIZE + 1] = { 0 };
+        BIO *biom = BIO_new(BIO_s_mem());
+        unsigned int cbrv = 0;
+
+        if (biom == NULL) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+        ossl_ech_status_print(biom, s, OSSL_ECHSTORE_ALL);
+        BIO_read(biom, pstr, OSSL_ECH_PBUF_SIZE);
+        cbrv = s->ext.ech.cb(&s->ssl, pstr);
+        BIO_free(biom);
+        if (cbrv != 1) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+/*
+ * do the HKDF for ECH acceptance checking
+ * md is the h/s hash
+ * for_hrr is 1 if we're doing a HRR
+ * hashval/hashlen is the transcript hash
+ * hoval is the output, with the ECH acceptance signal
+ * return 1 for good, 0 for error
+ */
+static int ech_hkdf_extract_wrap(SSL_CONNECTION *s, EVP_MD *md, int for_hrr,
+                                 unsigned char *hashval, size_t hashlen,
+                                 unsigned char hoval[OSSL_ECH_SIGNAL_LEN])
+{
+    int rv = 0;
+    unsigned char notsecret[EVP_MAX_MD_SIZE], zeros[EVP_MAX_MD_SIZE];
+    size_t retlen = 0, labellen = 0;
+    EVP_PKEY_CTX *pctx = NULL;
+    const char *label = NULL;
+    unsigned char *p = NULL;
+
+    if (for_hrr == 1) {
+        label = OSSL_ECH_HRR_CONFIRM_STRING;
+        labellen = sizeof(OSSL_ECH_HRR_CONFIRM_STRING) - 1;
+    } else {
+        label = OSSL_ECH_ACCEPT_CONFIRM_STRING;
+        labellen = sizeof(OSSL_ECH_ACCEPT_CONFIRM_STRING) - 1;
+    }
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ossl_ech_pbuf("cc: label", (unsigned char *)label, labellen);
+# endif
+    memset(zeros, 0, EVP_MAX_MD_SIZE);
+    /* We don't seem to have an hkdf-extract that's exposed by libcrypto */
+    pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+    if (pctx == NULL
+        || EVP_PKEY_derive_init(pctx) != 1
+        || EVP_PKEY_CTX_hkdf_mode(pctx,
+                                  EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY) != 1
+        || EVP_PKEY_CTX_set_hkdf_md(pctx, md) != 1) {
```

Comment:
I prefer `<= 0` in these comparisons, but OK to keep as is.

Reply:
if it's ok, I'll leave as-is (saves me refreshing that bit of my brain:-)